rest_framework_json_api/filters/filter.py

```diff
@@ -0,0 +1,135 @@
+        # and http://jsonapi.org/format/#document-member-names-reserved-characters for reserved
+        # characters (for use in paths, lists or as delimiters).
+        # regex `\w` matches [a-zA-Z0-9_].
+        # TODO: U+0080 and above allowed but not recommended. Leave them out for now. Fix later?
+        #       Also, ' ' (space) is allowed within a member name but not recommended.
+        filter_regex = re.compile(r'^filter(?P<ldelim>\[?)(?P<assoc>[\w\.\-]*)(?P<rdelim>\]?$)')
+
+        def validate_filter(self, keys, filterset_class):
+            for k in keys:
+                if ((not filterset_class) or (k not in filterset_class.base_filters)):
+                    raise ValidationError("invalid filter[{}]".format(k))
+
+        def get_filterset(self, request, queryset, view):
+            """
+            Sometimes there's no filterset_class defined yet the client still
+            requests a filter. Make sure they see an error too. This means
+            we have to get_filterset_kwargs() even if there's no filterset_class.
+
+            TODO: .base_filters vs. .filters attr (not always present)
+            """
+            filterset_class = self.get_filterset_class(view, queryset)
+            kwargs = self.get_filterset_kwargs(request, queryset, view)
+            self.validate_filter(self.filter_keys, filterset_class)
+            if filterset_class is None:
+                return None
+            return filterset_class(**kwargs)
+
+        def get_filterset_kwargs(self, request, queryset, view):
+            """
+            Turns filter[<field>]=<value> into <field>=<value> which is what
+            DjangoFilterBackend expects
+            """
+            self.filter_keys = []
+            # rewrite filter[field] query params to make DjangoFilterBackend work.
+            data = request.query_params.copy()
+            for qp, val in data.items():
+                m = self.filter_regex.match(qp)
+                if m and (not m.groupdict()['assoc'] or
+                          m.groupdict()['ldelim'] != '[' or m.groupdict()['rdelim'] != ']'):
+                    raise ValidationError("invalid filter: {}".format(qp))
+                if m and qp != self.search_param:
+                    if not val:
+                        raise ValidationError("missing {} test value".format(qp))
+                    # convert jsonapi relationship path to Django ORM's __ notation
+                    key = m.groupdict()['assoc'].replace('.', '__')
+                    # undo JSON_API_FORMAT_FIELD_NAMES conversion:
+                    key = format_value(key, 'underscore')
+                    data[key] = val
+                    self.filter_keys.append(key)
+                    del data[qp]
+            return {
+                'data': data,
+                'queryset': queryset,
+                'request': request,
+            }
+
+        def filter_queryset(self, request, queryset, view):
+            """
```

Comment:
hmm.... django-filters 2.0 does quiet a bit more here than 1.1. Not sure what errors this could cause.

Question is should we rather say we only support django-filters 2.0? This will also mean Python 2.7 users can not use it. But django-filters has done the same so we would align with upstream. And it makes our life certainly easier and we do not introduce potential errors with changes to support backwards compatibility.

What do you think?

Reply:
I think as long as upstream DRF and Django support py27 we should as well. django-filter has not completely dropped py27 yet; they've documented it [here in the README](https://github.com/carltongibson/django-filter/blame/master/README.rst#L27): 
> If you need to support Python 2.7 use the version 1.1 release.

All the unit tests pass for both versions. That tells me there might be more advanced usage in 2.0 that I'm not testing for, but it seems pretty solid given the number of tests.